nums/core/array/application.py

```diff
@@ -332,28 +335,67 @@ def diag(self, X: BlockArray) -> BlockArray:
                 syskwargs = {"grid_entry": grid_entry, "grid_shape": grid.grid_shape}
                 if np.all(np.diff(grid_entry) == 0):
                     # This is a diagonal block.
-                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid,
+                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid, [],
                                                                syskwargs=syskwargs)
                 else:
                     rarr.blocks[grid_entry].oid = self.cm.new_block("zeros",
                                                                     grid_entry,
                                                                     grid_meta,
                                                                     syskwargs=syskwargs)
         elif len(X.shape) == 2:
-            assert X.shape[0] == X.shape[1], "X must be a square array."
-            assert X.block_shape[0] == X.block_shape[1], "block_shape must be square."
-            shape = X.shape[0],
-            block_shape = X.block_shape[0],
+            shape = min(X.shape),
+            block_shape = min(X.block_shape),
             grid = ArrayGrid(shape, block_shape, X.dtype.__name__)
             rarr = BlockArray(grid, self.cm)
-            for grid_entry in X.grid.get_entry_iterator():
-                out_grid_entry = grid_entry[:1]
-                out_grid_shape = grid.grid_shape[:1]
-                syskwargs = {"grid_entry": out_grid_entry, "grid_shape": out_grid_shape}
-                if np.all(np.diff(grid_entry) == 0):
-                    # This is a diagonal block.
-                    rarr.blocks[out_grid_entry].oid = self.cm.diag(X.blocks[grid_entry].oid,
-                                                                   syskwargs=syskwargs)
+            block_i, block_j, element_i, element_j = 0, 0, 0, 0
+            count, total_elements = 1, min(shape)
+            block, diagonal, block_no, current_block_length = X.blocks[(0, 0)], OrderedDict(), 1, 0
+            while count <= total_elements:
+                if element_i > block.shape[0] - 1:
+                    block_i = block_i + 1
+                    element_i = 0
+                if element_j > block.shape[1] - 1:
+                    block_j = block_j + 1
+                    element_j = 0
+                block = X.blocks[(block_i, block_j)]
+                while True:
```

Comment:
We cannot. Because even if we aren't doing multiple input blocks => one output block, we need the inner loop for the elements we need from a particular block. I think we already discussed about the importance of the inner loop in our last meeting. Probably the additions related to the multiple input blocks stuff makes it look like if it isn't needed.